### DISCUSSION TITLE
useMemo and useCallback execute once during server-side rendering?

### DISCUSSION THREAD

**Discussion by catgrandfa** [vercel/next.js#55316] [Help]
### Link to the code that reproduces this issue or a replay of the bug

https://codesandbox.io/p/sandbox/festive-forest-2xwykd

### To Reproduce

` const testMemo = useMemo(() => {
    console.log("testMemotestMemotestMemotestMemotestMemotestMemo");
    return JSON.stringify(props);
  }, [props]);`
  
<img width="914" alt="image" src="https://github.com/vercel/next.js/assets/53386949/3a32d124-8c6a-4c9c-a5a2-aaa6d54eb081">

### Current vs. Expected behavior

When I use the useMemo and useCallback hooks, they seem to execute once during server-side rendering, causing an error due to a mismatch between the server and client results. However, based on my understanding from the Next.js documentation, hooks are supposed to run only on the client-side. Why is this happening?

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.6.0: Wed Jul  5 22:22:05 PDT 2023; root:xnu-8796.141.3~6/RELEASE_ARM64_T6000
    Binaries:
      Node: 16.20.1
      npm: 8.19.4
      Yarn: 1.22.19
      pnpm: 8.6.2
    Relevant Packages:
      next: 13.4.19
      eslint-config-next: 13.4.9
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.6
    Next.js Config:
      output: N/A
```

### Which area(s) are affected? (Select all that apply)

App Router

**Comment by icyJoseph** [comment]
> However, based on my understanding from the Next.js documentation, hooks are supposed to run only on the client-side. Why is this happening?

Well that's not correct. Not sure how you got to that conclusion. 

- https://github.com/reactwg/server-components/discussions/4

Client components are indeed rendered on the server. The key difference is that, server components, only render server side.

> Client components are the same exact components we have had up till now.

```jsx
'use client';

export const Foo = () => {
   const foo = useMemo(() => "foo", [])
   return <div>{foo}</div>
}
```

So in order to SSR the Foo client component, we need to calculate `foo`, which is why the useMemo is executed.

Maybe you want to take a look at this section: https://demystifying-rsc.vercel.app/client-components/no-ssr/, and possibly the entire study.

**Comment by catgrandfa** [comment]
Thank you for your reply. I will try to understand this.

2023年9月13日 15:55，Joseph ***@***.***> 写道：



However, based on my understanding from the Next.js documentation, hooks are supposed to run only on the client-side. Why is this happening?

Well that's not correct. Not sure how you got to that conclusion.

  *   reactwg/server-components#4<https://github.com/reactwg/server-components/discussions/4>

Client components are indeed rendered on the server. The key difference is that, server components, only render server side.

'use client';

export const Foo = () => {
   const foo = useMemo(() => "foo", [])
   return <div>{foo}</div>
}

So in order to SSR the Foo client component, we need to calculate foo, which is why the useMemo is executed.

**Comment by kirim-Lee** [comment]
when excute useMemo or useCallback function in server once, window is undefined so it should use server action function, but server function like cookies() 'next/headers' also doesn't not defined. so it occures error. 
```
// getContentUrl reads cookie (when server function that uses cookies(), when client function that uses window.cookie. 
// but server function cookies() is not undefined)
const url = useMemo(() => getContentUrl(content), [content])
```
